<a href="https://colab.research.google.com/github/ismailsangu/IE423-TASK/blob/main/Task_9.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Pretrained Models:
#Can you speed up your efforts using pretrained models?

# Transfomers

Pretrained chains of models that perform specific tasks

## Initialize

In [1]:
!pip install transformers

In [2]:
# Use pipelines to access pre-trained models
from transformers import pipeline

## Sentiment Analysis

In [3]:
# Model
mdlSnt = pipeline('sentiment-analysis')

# Parameters
#prmStatement = 'We are happy to go on vacation this spring break.'
prmStatement = ['It is nice to have you here in the US.  I heard the weather is terrible in the UK.',
        'I am struggling with stress']
# Predict
results=mdlSnt(prmStatement)
# Display results
for statement, result in zip(prmStatement, results):
    label = result['label']
    score = result['score']
    print(f"Statement: \"{statement}\"")
    print(f"Sentiment: {label} (Score: {score:.4f})\n")



No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision af0f99b (https://huggingface.co/distilbert/distilbert-base-uncased-finetuned-sst-2-english).
Using a pipeline without specifying a model name and revision in production is not recommended.
/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/629 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/268M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

Statement: "It is nice to have you here in the US.  I heard the weather is terrible in the UK."
Sentiment: POSITIVE (Score: 0.8163)

Statement: "I am struggling with stress"
Sentiment: NEGATIVE (Score: 0.9938)



Sentiment Analysis can be used in ordering choices.

In [4]:
from transformers import pipeline

# Load the sentiment-analysis model
mdlSnt = pipeline('sentiment-analysis')

# List of statements
prmStatement = [
    'I am student in Bilkent University','I am student in Middle East Technical University','I am student in Koç University','I am student in Boğaziçi University','I am student in Oxford University','I am student in MIT','I am student in Sabancı University'
,'I am student in Abdullah Gül University']

# Get sentiment analysis results
results = mdlSnt(prmStatement)

# Combine statements with their results
combined_results = list(zip(prmStatement, results))

# Sort by sentiment score:
# - Positive sentiments: sorted by score in descending order
# - Negative sentiments: sorted by score in ascending order
combined_results_sorted = sorted(
    combined_results,
    key=lambda x: (x[1]['label'] == 'NEGATIVE', -x[1]['score']) if x[1]['label'] == 'POSITIVE' else (x[1]['label'] == 'NEGATIVE', x[1]['score'])
)

# Display sorted results with formatting
print("\nUNIVERSITY PREFERENCE:\n")
for index, (statement, result) in enumerate(combined_results_sorted, start=1):
    label = result['label']
    score = result['score']
    print(f"{index}. \"{statement}\"")
    print(f"   Sentiment: {label}")
    print(f"   Score: {score:.4f}\n")

No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision af0f99b (https://huggingface.co/distilbert/distilbert-base-uncased-finetuned-sst-2-english).
Using a pipeline without specifying a model name and revision in production is not recommended.



UNIVERSITY PREFERENCE:

1. "I am student in Abdullah Gül University"
   Sentiment: POSITIVE
   Score: 0.9205

2. "I am student in Koç University"
   Sentiment: POSITIVE
   Score: 0.5826

3. "I am student in Sabancı University"
   Sentiment: POSITIVE
   Score: 0.5111

4. "I am student in Boğaziçi University"
   Sentiment: POSITIVE
   Score: 0.5061

5. "I am student in Oxford University"
   Sentiment: POSITIVE
   Score: 0.5010

6. "I am student in MIT"
   Sentiment: NEGATIVE
   Score: 0.7005

7. "I am student in Middle East Technical University"
   Sentiment: NEGATIVE
   Score: 0.8394

8. "I am student in Bilkent University"
   Sentiment: NEGATIVE
   Score: 0.9941



Hmmmm, there are interesting results.

In [5]:
from transformers import pipeline

# Load the sentiment-analysis model
mdlSnt = pipeline('sentiment-analysis')

# List of statements
prmStatement = [
    'I am mechanical engineer','I am industrial engineer','I am computer engineer','I am electrical and electronics engineer'
]

# Get sentiment analysis results
results = mdlSnt(prmStatement)

# Combine statements with their results
combined_results = list(zip(prmStatement, results))

# Sort by sentiment score:
# - Positive sentiments: sorted by score in descending order
# - Negative sentiments: sorted by score in ascending order
combined_results_sorted = sorted(
    combined_results,
    key=lambda x: (x[1]['label'] == 'NEGATIVE', -x[1]['score']) if x[1]['label'] == 'POSITIVE' else (x[1]['label'] == 'NEGATIVE', x[1]['score'])
)

# Display sorted results with formatting
print("\nENGINEERING DEPARTMENT PREFERENCE:\n")
for index, (statement, result) in enumerate(combined_results_sorted, start=1):
    label = result['label']
    score = result['score']
    print(f"{index}. \"{statement}\"")
    print(f"   Sentiment: {label}")
    print(f"   Score: {score:.4f}\n")

No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision af0f99b (https://huggingface.co/distilbert/distilbert-base-uncased-finetuned-sst-2-english).
Using a pipeline without specifying a model name and revision in production is not recommended.



ENGINEERING DEPARTMENT PREFERENCE:

1. "I am electrical and electronics engineer"
   Sentiment: POSITIVE
   Score: 0.9932

2. "I am mechanical engineer"
   Sentiment: POSITIVE
   Score: 0.9851

3. "I am industrial engineer"
   Sentiment: POSITIVE
   Score: 0.9844

4. "I am computer engineer"
   Sentiment: POSITIVE
   Score: 0.9600



Can perform sentiment analysis in other languages as well using `model = 'nlptown/bert-base-multilingual-uncased-sentiment'`

In [19]:
from transformers import pipeline

# Load the sentiment-analysis model
mdlSnt_Turkish = pipeline('sentiment-analysis', model='savasy/bert-base-turkish-sentiment-cased')

# List of statements
prmStatement_Turkish = [
    'Ben makine mühendisiyim','Ben endüstri mühendisiyim','Ben bilgisayar mühendisiyim','Ben elektrik ve elektronik mühendisiym'
]

# Get sentiment analysis results
results_Turkish = mdlSnt_Turkish(prmStatement_Turkish)

# Combine statements with their results
combined_results_Turkish = list(zip(prmStatement_Turkish, results_Turkish))

# Sort by sentiment score:
# - Positive sentiments: sorted by score in descending order
# - Negative sentiments: sorted by score in ascending order
combined_results_sorted_Turkish = sorted(
    combined_results_Turkish,
    key=lambda x: (x[1]['label'] == 'NEGATIVE', -x[1]['score']) if x[1]['label'] == 'POSITIVE' else (x[1]['label'] == 'NEGATIVE', x[1]['score'])
)

# Display sorted results with formatting
print("\nMühendislik departman Tercihi:\n")
for index, (statement, result) in enumerate(combined_results_sorted_Turkish, start=1):
    label = result['label']
    score = result['score']
    print(f"{index}. \"{statement}\"")
    print(f"   Sentiment: {label}")
    print(f"   Score: {score:.4f}\n")


Mühendislik departman Tercihi:

1. "Ben elektrik ve elektronik mühendisiym"
   Sentiment: negative
   Score: 0.8491

2. "Ben makine mühendisiyim"
   Sentiment: negative
   Score: 0.9594

3. "Ben bilgisayar mühendisiyim"
   Sentiment: negative
   Score: 0.9647

4. "Ben endüstri mühendisiyim"
   Sentiment: negative
   Score: 0.9782



It also works well in Turkish but interestingly they are all described as a negative and have different score from English statments.

## Question Answering

### Example 1

In [7]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [8]:
# Model
mdlQa = pipeline("question-answering")

# Parameters
f = open("/content/drive/MyDrive/Data_Sets/HenryFord.txt", "r")
context = f.read()
#context = Henry Ford's Bioraphy
question = "What was the Henry Ford's job?"

# Predict
mdlQa(question = question, context = context)

No model was supplied, defaulted to distilbert/distilbert-base-cased-distilled-squad and revision 626af31 (https://huggingface.co/distilbert/distilbert-base-cased-distilled-squad).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/473 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/261M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/49.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/213k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/436k [00:00<?, ?B/s]

{'score': 0.6220643520355225,
 'start': 1410,
 'end': 1432,
 'answer': 'builder of automobiles'}

It is correct answer.

In [9]:
#context = Henry Ford's Bioraphy
question = "When was the Ford Motor Company Founded?"

# Predict
mdlQa(question = question, context = context)

{'score': 0.9688721299171448,
 'start': 123,
 'end': 136,
 'answer': 'July 30, 1863'}

It is not correct. The answer is birtday date of Henry Ford, not Ford Motor Company foundation date.The model takes <font color='#378'>"Henry Ford, founder of Ford Motor Company, was born in Springwells Township, Wayne County, Michigan, on July 30, 1863."<font color='#999'> as an answer instead of <font color='#472'>"On June 16, 1903, Henry and 12 others invested $28,000 and created Ford Motor Company."

In first insight, the reason of mistake seems like pickin up closest data behavior of the model but when we change the place of these above two phrases, the answer does not change. It may be solved by adding more context into model.

In [10]:
#context = Henry Ford's Bioraphy
question = "When was the Ford Motor Company Founded by twelve investers?"

# Predict
mdlQa(question = question, context = context)

{'score': 0.9496527314186096,
 'start': 123,
 'end': 136,
 'answer': 'July 30, 1863'}

In [11]:
#context = Henry Ford's Bioraphy
question = "When was the Ford Motor Company Founded by 12 investers?"

# Predict
mdlQa(question = question, context = context)

{'score': 0.9513087868690491,
 'start': 2665,
 'end': 2678,
 'answer': 'June 16, 1903'}

While adding more detail to the question gives still inaccurate answer, adding key word '12' gives accurate result. Hence, it can be claimed that the effect of keyword on model is more than effect of more detail and more text.